### PR TITLE
reimplement arg passthrough for clippy-driver

### PR DESCRIPTION
**Pull request by yaahc**
Prior to this change, the old cargo subcommand version of `cargo clippy`
had a feature to pass trailing args down to clippy-driver but when this
the subcommand was reimplemented inside of cargo this feature was left
out accidentally.

This change readds the feature to to capture all args after a trailing
-- and forward them down to clippy-driver via an env variable.

### CODE REVIEW

**Comment by yaahallo** [comment]
One thing I want to mention is that the way that clippy passes the args down is inconsistent with how we did it for `cargo fix --clippy`

For cargo fix clippy we take the string associated with the `--clippy` arg if any is specified `--clippy="--deny clippy::all"` where as for `cargo clippy` we take anything after a trailing `--`, so `cargo clippy -- --deny clippy::all`.

Also this is implemented differently, clippy expects its args to be passed via a special env variable with the a big `__CLIPPY_HACKERY__` appended to the end of each arg to help with arg splitting I assume. With `cargo fix --clippy` we serialize the vec of split args as json and deserialize to preserve word splitting.

I'd like to see the visible portion of how the args are specified become consistent, at the very least. So at a minimum I'd like trailing `-- [args]...` added to fix which can be used for passthrough to clippy and possibly rustc to configure lints. I believe clippy's approach is the better one in this case.

As for the implementation details I'd love for it to be consistent but clippy doesn't currently bring in serde_json, and so I'm not sure if its worth going through the effort here to make it consistent or not. @Manishearth input is welcome.

File: src/bin/cargo/commands/clippy.rs [hunk 1]
```diff
@@ -69,7 +70,15 @@ pub fn exec(config: &mut Config, args: &ArgMatches<'_>) -> CliResult {
...
+    if let Some(old_args) = args.values_of("args") {
+        let clippy_args: String = old_args
+            .map(|arg| format!("{}__CLIPPY_HACKERY__", arg))
+            .collect();
+        wrapper.env("CLIPPY_ARGS", clippy_args);
+    }
+
     compile_opts.build_config.primary_unit_rustc = Some(wrapper);
 
     ops::compile(&ws, &compile_opts)?;
```
Comment (yaahallo) [inline src/bin/cargo/commands/clippy.rs]: I'll add a test for this in a hot minute

File: src/bin/cargo/commands/clippy.rs [hunk 2]
```diff
@@ -69,7 +70,15 @@ pub fn exec(config: &mut Config, args: &ArgMatches<'_>) -> CliResult {
         .into());
     }
 
-    let wrapper = util::process(util::config::clippy_driver());
+    let mut wrapper = util::process(util::config::clippy_driver());
+
+    if let Some(old_args) = args.values_of("args") {
+        let clippy_args: String = old_args
```
Comment (yaahallo) [inline src/bin/cargo/commands/clippy.rs]: I'll add a test for this in a bit.

**Comment by Manishearth** [comment]
Can we not use CLIPPY_HACKERY? Part of the goal here is that such hacks aren't needed anymore.

We can instead use another trailing -- on clippy. 

Also, it's totally okay if `cargo clippy-preview` has a different interface than the old cargo clippy. The hope is to improve things.

**Comment by yaahallo** [comment]
I'm stupid, we can definitely not use clippy_hackery, we dont even need to use it here because we already have the processbuilder, no resplitting is required.

**Comment by yaahallo** [comment]
dont approve it yet pls, still working on resolving the clippy_hackery

**Comment by yaahallo** [comment]
Okay, I removed `__CLIPPY_HACKERY__` and did the tweeks necessary to have clippy run as rustc instead of as a wrapper so it all works together nicely. I think the stuff to fix the `-- [args]...` for fix should be done as an independent commit so all this needs is a test and code review + pass CI.

**Comment by yaahallo** [comment]
right on schedule bors, ty bby

**Comment by ehuss** [comment]
Thanks!

Instead of introducing a new enum, I think we can just make the rustc argument explicit for `fix`.  That is:
- Call something like this in `ops::fix`:
  ```rust
  let rustc = opts.compile_opts.config.load_global_rustc(Some(ws))?;
  wrapper.arg(&rustc.path);
  ```
- Remove the call to `map` [here](https://github.com/rust-lang/cargo/blob/137a36f07004c33eaf742b85c04abe296aa4aa17/src/cargo/core/compiler/compilation.rs#L82) so it no longer automatically adds the argument.

**Comment by yaahc** [comment]
Okay I think it's good, rereading the code it looks super clean, TY again @ehuss for always providing excellent guidance.

**Comment by ehuss** [comment]
Thanks!

@bors r+

**Pull request merged**